Sprint3/step1

3hunnaDev opened the pull request.

3hunnaDev committed:
update API base URL and add react-router-dom dependency

3hunnaDev committed:
Перевёл приложение на React Router, добавил модалки/страницы ингредиентов и полноценный модуль авторизации (логин, регистрация, профиль, восстановление пароля) с защищёнными маршрутами и проверкой токенов при оформлении заказа

3hunnaDev committed:
feat: Implement user authentication and profile management

- Add profile management component with editable fields for name, email, and password.
- Create registration and login components with form validation.
- Implement password reset functionality with token verification.
- Add ingredient details modal for displaying ingredient information.
- Create protected route component to manage access based on authentication status.
- Develop shared text input component for consistent input handling.
- Introduce custom hook for fetching ingredient details.
- Set up Redux store for authentication state management with thunks for async actions.
- Implement cookie utility functions for managing access and refresh tokens.
- Add validation utilities for email and password fields.
- Create pages for forgot password, home, ingredient details, login, not found, orders, profile, register, and reset password.
- Style components with CSS modules for better layout and design.

gennady-bars commented on src/components/app-authorization/authorization-login/login.tsx:
```diff
@@ -0,0 +1,106 @@
+    (state: RootState) => state.auth
+  );
+  const [form, setForm] = useState({ email: "", password: "" });
+
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/components/app-authorization/authorization-forgot-password/forgot-password.tsx:
```diff
@@ -0,0 +1,89 @@
+import React, { useCallback, useEffect, useState } from "react";
+import { Link, useNavigate } from "react-router-dom";
+import { useDispatch, useSelector } from "react-redux";
```
**Можно лучше**

`import { useDispatch, useSelector } from 'react-redux';` это нужно убрать из проекта.  Хуки `useDispatch, useSelector` нужно типизировать **1 раз** в отдельном файле и заменить их во всем проекте на свои, а не из библиотеки `react-redux` .  Вот ссылка на документацию по их типизации https://redux.js.org/usage/usage-with-typescript#define-typed-hooks

gennady-bars commented on src/components/app-authorization/authorization-forgot-password/forgot-password.tsx:
```diff
@@ -0,0 +1,89 @@
+const ForgotPassword: React.FC = () => {
+  const dispatch = useDispatch<AppDispatch>();
+  const navigate = useNavigate();
+  const { status, errors } = useSelector((state: RootState) => state.auth);
```
**Можно лучше**

`(state: RootState)` уже не нужно будет типизировать вручную стейт, когда замените везде хук `useSelector`. Нужно будет удалить дублирование типизации

gennady-bars commented on src/components/app-authorization/authorization-profile/profile.tsx:
```diff
@@ -0,0 +1,257 @@
+  const renderProfileContent = () => (
+    <form
+      className={styles.formWrapper}
+      onSubmit={(event) => event.preventDefault()}
```
Вот тут в теге `form` нужно указать обработчик сабмита с отправкой запроса `dispatch(updateUserProfile(payload));`

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично сделана логика отображения попапа и страницы  под ним
* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку


но есть некоторые недочеты:



* Почему-то ошибка при отправке заказа на сервер. Возможно, какие-то проблемы на самом сервере в момент ревью, но лучше перепроверить
* Вместо `AppHeaderTab` нужно использовать компонент `NavLink` из библиотеки роутинга, который может определять активную страницу. По заданию нужно **подсвечивать** в хэдере ту ссылку, где пользователь сейчас находится.
* Не стоит в `handleFieldBlur` делать запрос на сервер для сохранения изменений в профиле. Это делается при сабмите формы через `Enter` или клике по кнопке сабмита **Сохранить**, которая вызовет событие `submit` для сабмита формы
* Если в профиле начать изменять данные в имени, то по макету в фигме должны появиться 2 кнопки: **Сохранить** и **Отменить** https://disk.yandex.ru/i/sMiYsVoomx-Ozw  . При отмене данные возвращаются к начальному состоянию в инпутах
* `dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.


**МОЖНО ЛУЧШЕ**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `.then` или после закрытия попапа с номером заказа, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа
* при клике на **логотип** обычно идет переход на **главную** страницу

3hunnaDev committed:
feat: Refactor authentication and profile management components to use custom hooks and improve code structure
fix: Update order creation API to include token in the request body
style: Adjust profile sidebar width and add new styles for notices and form actions
chore: Remove unused navigation reducer and related components, streamline app header navigation
add: Implement loading state for order submission with a new loader component

3hunnaDev committed:
feat: Add keyboard navigation to profile form for canceling edits with Escape key

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
